src/platform/packages/shared/kbn-esql-ast/src/commands_registry/commands/join/autocomplete.ts

```diff
@@ -117,31 +126,63 @@ export async function autocomplete(
       return [suggestion];
     }
 
-    case 'after_on': {
-      return await suggestFields(
-        innerText,
-        command,
-        callbacks?.getByType,
-        callbacks?.getColumnsForQuery,
+    case 'on_expression': {
+      const joinCommand = command as ESQLAstJoinCommand;
+      const expressionRoot = position.expression;
+
+      // Create enriched getByType that includes lookup fields
+      const enrichedGetByType = await createEnrichedGetByType(
+        callbacks?.getByType ?? (() => Promise.resolve([])),
+        joinCommand,
+        (callbacks?.getColumnsForQuery ?? (() => Promise.resolve([]))) as (
+          query: string
+        ) => Promise<ESQLFieldWithMetadata[]>,
         context
       );
-    }
 
-    case 'condition': {
-      if (/(?<!\,)\s+$/.test(innerText)) {
-        // this trailing whitespace was not proceeded by a comma
-        return [commaCompleteItem, pipeCompleteItem];
-      }
+      // Create enriched context that includes lookup fields in columns map
+      const enrichedContext = await createEnrichedContext(
+        context,
+        joinCommand,
+        (callbacks?.getColumnsForQuery ?? (() => Promise.resolve([]))) as (
+          query: string
+        ) => Promise<ESQLFieldWithMetadata[]>
+      );
 
-      const fields = await suggestFields(
-        innerText,
+      const suggestions = await suggestForExpression({
+        query,
+        expressionRoot,
         command,
-        callbacks?.getByType,
-        callbacks?.getColumnsForQuery,
-        context
-      );
+        cursorPosition,
+        location: Location.JOIN,
+        context: enrichedContext,
+        callbacks: {
+          ...callbacks,
+          getByType: enrichedGetByType,
+        },
+        options: {
+          preferredExpressionType: 'boolean',
+        },
+      });
+
+      // Filter out AS operator - it's not valid in boolean expressions
+      const filteredSuggestions = suggestions.filter(({ label }) => label !== 'AS');
```

Comment:
`AS `uses` location.JOIN` but should not be suggested within ON.
A manual filter is sufficient for this specific case, instead of creating a new` Location.JOIN_ON`